Love for remote debugging

deivid-rodriguez opened the pull request:
Fixes #82.
Fixes #141.
Fixes #239.
Fixes #274.
Fixes #302.

deivid-rodriguez commented:
Summoning @MSP-Greg in case you're interested. I've been working on remote debugging support. Not ready yet but looking good. However, I'm getting some failures on Windows and I don't have a Windows machine to debug. In case you would like to help out, I suspect it's something around setting binary mode in streams, but not sure. Regards!

MSP-Greg commented:
@deivid-rodriguez

I'll have a look at it, might not be able to seriously do so until Saturday...  Thanks, Greg

deivid-rodriguez commented:
That's great, thanks so much!

MSP-Greg commented:
@deivid-rodriguez

Solution found.  See  [commit in my fork](https://github.com/MSP-Greg/byebug/commit/9ef60313c61b58c97c72c6f18b7b528c6021e42d), [Appveyor results](https://ci.appveyor.com/project/MSP-Greg/byebug)

Probably not what you (or I) expected.  Hope this helps, Greg

MSP-Greg commented:
Added another patch to swap the fork for Thread.  See [commit in my fork](https://github.com/MSP-Greg/byebug/commit/ed8205b8261cf9ffe175f7d785724aa4cd68164f).  [Appveyor results](https://ci.appveyor.com/project/MSP-Greg/byebug/build/2).

Appveyor runs with dual cores, Travis has four.  The `sleep(windows? ? 2 : 1)` line may need windows increased to 2.5 or 3, not sure.  The first build I ran stopped with a very strange log, second build passed...

deivid-rodriguez commented:
@MSP-Greg Thanks so much! Your fix makes a lot of sense since only `byebug`'s executable [sets up the LOAD_PATH](https://github.com/deivid-rodriguez/byebug/blob/1b55351e045ed67cd3e87493a5996ccb77a77f4c/exe/byebug#L4-L5) but it can't be used directly on Windows!

Also thanks for letting me know about the ruby-core issue, I commented in there! And also thanks for fixing those tests to not use `fork`.

I'll either give you access to this branch to add your changes, or cherry-pick them myself!

MSP-Greg commented:
@deivid-rodriguez

> Also thanks for letting me know about the ruby-core issue, I commented in there!

Thank you.  I keep an eye on it (and the ruby-core Travis & Appveyor), and I knew you could address the issue(s) well.

I'm also sensitive to issues that involve trunk, If someone is concerned about a given repo/gem functioning with trunk, they're welcome to fork and run their own CI, since many 'mature' repos/gems may not have frequent commits...

> I'll either give you access to this branch to add your changes, or cherry-pick them myself!

Whatever is easiest.

Finally, I reverted 'Binmode?' with two commits, both passed.

Thanks, Greg

MSP-Greg commented:
@deivid-rodriguez

Would you prefer that I delete or revert the `Binmode?` commit (since it is the last one)?

deivid-rodriguez commented:
Delete :+1:

MSP-Greg commented:
Sometimes Appveyor...  Had this issue last night, where I changed `sleep 1` to `sleep(windows? ? 2 : 1)`.

Just upped it to 3, no change.  I'm going to check/fix it on my fork.  I must be missing something....

FYI - from all my testing on Appveyor (100's of trunk builds, rubygems builds, etc), I've seen lots of intermittent issues...

MSP-Greg commented:
@deivid-rodriguez

I found what the issue was with my fork passing the one test.  I dl'd a diff of the fork, and (oddly) one line was different.  I even checked my browser history, and timestamps matched, so it was what GitHub had.  The line [`def write_program(code)`](https://github.com/deivid-rodriguez/byebug/blob/fix/remote_debugging/test/remote_test.rb#L138) was changed to `def write_program(program)`.

Anyway, fixed that, I think I've got this working, juts a few minutes.  The change is to [`RemoteInterface#readline`](https://github.com/deivid-rodriguez/byebug/blob/7bddb3e688d1a203f962c2b3570974a50abc2771/lib/byebug/interfaces/remote_interface.rb#L41-L48)

to this:

```ruby
def readline(prompt)
  puts(prompt)

  result = (input.gets || "continue")
  result.chomp
rescue Errno::EPIPE, Errno::ECONNABORTED
  "continue"
end
```

`Errno::EPIPE` is raised by trunk, `Errno::ECONNABORTED` is raised by 2.4 and lower.

Thoughts?  trying on my fork shortly...  Thanks, Greg

deivid-rodriguez commented:
> I found what the issue was with my fork passing the one test. I dl'd a diff of the fork, and (oddly) one line was different. I even checked my browser history, and timestamps matched, so it was what GitHub had. The line def write_program(code) was changed to def write_program(program).

Yeah, sorry! I realized that was incorrect and fixed it here, leaving your fork up to date indeed. Feel free to just work on this same branch instead of your fork, I'll be carefull with force-pushing :)

Your fix looks good, but I think the `rescue` should be moved up to the `confirm` and `read_command` methods (which internally call `readline`), similarly to `puts` and `print`. Otherwise I think the exception could be raised outside of the `rescue` block you're adding. What do you think?

MSP-Greg commented:
> Yeah, sorry!

That was before I got changed to 'collaborator', so I grabbed the changes and was working just in my fork in a different new branch.  IOW, no problem.

> I'll be carefull with force-pushing :)

I do that alot also.  I'll try to restrict it to things I do **before** I push here.

> I think the rescue should be moved up to the confirm and read_command methods (which internally call readline), similarly to puts and print. Otherwise I think the exception could be raised outside of the rescue block you're adding. What do you think?

Well, it (as listed) just [passed](https://ci.appveyor.com/project/MSP-Greg/byebug/build/7)  on my fork with Appveyor.  The commit is [here](https://github.com/MSP-Greg/byebug/commit/ce1acad6a4122517fadf63cba8a6eb7614ae27a5).

I haven't reviewed the remote code enough to know if a particular error should trigger a 'is the client  dead?' check,..

Lastly, can we please remove the 'stop on failure' from Appveyor?  I normally do all of my work with trunk.  I was passing locally, pushed to Appveyor, and the build stopped on 2.3 (which lead to my adding `Errno::ECONNABORTED`).

deivid-rodriguez commented:
> Well, it (as listed) just passed on my fork with Appveyor. The commit is here.

Yeah, we can't know where the exception will be raised, and most of the times it will happen under the `rescue` block, but my alternative should be safer I think.

> Lastly, can we please remove the 'stop on failure' from Appveyor? I normally do all of my work with trunk. I was passing locally, pushed to Appveyor, and the build stopped on 2.3 (which lead to my adding Errno::ECONNABORTED).

Ok, since you are the only one here working on Windows, let's do whatever works best for you :)

MSP-Greg commented:
> Yeah, we can't know where the exception will be raised, and most of the times it will happen under the rescue block, but my alternative should be safer I think.

I'll defer to you.  I'm just starting at the tests and moving inward.  You know the whole thing.

Changed. Test [passed](https://ci.appveyor.com/project/MSP-Greg/byebug/build/8).  Commit is [here](https://github.com/MSP-Greg/byebug/commit/32fe8eed5e82b2a6ada89919102527782d31908f).  I changed the `readline` code to remove the `result` variable.  Old habit, I prefer to never use local variables if not needed, because 6 months from now, I'd feel a need to double check if `result` was an attribute I'd forgotten about.  I think that made sense....

Let me know if it's good, and I'll push it.  then I'll move on to #416...

deivid-rodriguez commented:
> Changed. Test passed. Commit is here.

Yep, that's what I meant. However, the `confirm` part should probably default to `false`, not `"continue"` since `super` returns a boolean. That said, that part is untested at the moment, so you can just leave it alone for now. So I'd say no change at all there, or rescue and return `false`, up to you!

> I changed the readline code to remove the result variable. Old habit, I prefer to never use local variables if not needed, because 6 months from now, I'd feel a need to double check if result was an attribute I'd forgotten about. I think that made sense..

Mmmm, not sure. I think local variables have preference over attributes, if you were to set an attribute, you'd have to do `self.result =`, I think? The new code is more concise but unnecessarily `chomps` "continue". So I'm "meh" on the change, but let's do it if you prefer it that way.

MSP-Greg commented:
> rescue and return false, up to you!

Sorry I missed that.  Changed.

>  unnecessarily chomps "continue".

Can't chomp nil though...

Pushing shortly.  Thanks.

deivid-rodriguez commented:
> Sorry I missed that. Changed.

:+1: 

> > unnecessarily chomps "continue".
>
> Can't chomp nil though...

I mean that, correctly but unnecessarily, calls `"continue".chomp` (which returns "continue" again) when `input.gets` is `nil`. I do like its simplicity though, let's do it! :+1:

MSP-Greg commented:
Cool. Sorry, I meant that `input.gets.chomp || "continue"` could error.

Thanks for cancelling Appveyor...

deivid-rodriguez commented:
> Cool. Sorry, I meant that input.gets.chomp || "continue" could error.

Yup, that could fail. It was you way or mine, yours works :)

> Thanks for cancelling Appveyor...

No problem.

I'll give this a final look tomorrow, hopefully I can add coverage for a couple more cases!

deivid-rodriguez commented:
@MSP-Greg I just rebased this PR!

I had also added a some more tests and refactorings to the remote debugging specs but you know what? It made it hang again. :rage:

That gave me a clue though to further isolate the bug / problem with `Open3` + `rb-readline`. Maybe it happens when you use threads inside `popen2e`'s block? One difference between the current state of this PR and the new tests I had added or what we are doing in #416 is that we currently don't do that here... Note that [`capture2e` implementation](https://github.com/ruby/ruby/blob/3d9453994c965abff8851d008a01b40153ffee77/lib/open3.rb#L348-L368), which we are using in #416 uses `Thread` inside `popen2e`'s block...

MSP-Greg commented:
@deivid-rodriguez

Just looked, the last AV run passed.  Things good now?

Re `Open3`, I've got the readline prompt appearing, but after that, it goes somewhere unknown.  I've tried with and without threads.   Of course, when I switch to `readline.so`, the code runs fine...

deivid-rodriguez commented:
Rebased and added new tests and refactors and the Windows build passed! :tada:

deivid-rodriguez commented on lib/byebug/interfaces/remote_interface.rb:
```diff
@@ -16,25 +16,35 @@ def initialize(socket)
+
+    def print(message)
+      super(message)
+    rescue Errno::EPIPE
```
@MSP-Greg Should we rescue `Errno::ECONNABORTED` here and in `puts` as well? You were seeing that exception on Windows, but only under some versions if I recall correctly?

MSP-Greg replied:
I don't recall, I'll have to check.

deivid-rodriguez replied:
You mentioned it [here](https://github.com/deivid-rodriguez/byebug/pull/406#issuecomment-355767858). I think I'll add `Errno::ECONNABORTED` in the other methods as well then.

Re `exe/byebug`, feel free to suggest any changes you want in a separate PR!

MSP-Greg replied:
> You mentioned it here.

Sorry, rather distracted with `non code` things yesterday evening.

Travis has had large Linux & MacOS backlogs recently (separate occasions).  Hope they get it sorted out...

deivid-rodriguez commented:
Let's do this!
